tensorflow/python/kernel_tests/nn_ops/cudnn_deterministic_base.py

```diff
@@ -71,98 +73,117 @@ def _random_data_op(self, shape):
     return constant_op.constant(
         2 * np.random.random_sample(shape) - 1, dtype=dtypes.float32)
 
-  def _random_out_op(self, in_shape, filter_shape, strides, padding):
+  def _random_out_op(self, in_shape, filter_shape, strides, padding, dilations):
     # Choosing not to use array_op.zeros() to prevent possible removal by
     # optimization
     in_op = self._random_data_op(in_shape)
     filter_op = self._random_data_op(filter_shape)
     # Use the forward op's shape-inference
-    conv_op = nn_ops.conv2d(in_op, filter_op, strides=strides, padding=padding)
+    conv_op = nn_ops.conv2d(
+        in_op, filter_op, strides=strides, padding=padding, dilations=dilations)
     out_shape = conv_op.get_shape()
     out_op = self._random_data_op(out_shape)
     return out_op
 
   def _assert_reproducible(self, operation):
-    with self.cached_session(force_gpu=True):
-      result_1 = self.evaluate(operation)
-      result_2 = self.evaluate(operation)
+    with test_util.force_gpu():
+      result_1 = operation()
+      result_2 = operation()
     self.assertAllEqual(result_1, result_2)
 
   # The default forward algorithm choice, when using cuDNN 7, does not support
   # the following layer configuration. This test case intends to confirm that
   # an alternative algorithm is selected. Note that, in cuDNN 7, all forward
   # algorithms are determnistic.
   @test_util.run_cuda_only
-  def testForward(self):
+  def testConvForwardDefaultAlgorithmChoice(self):
     in_shape = LayerShapeNCDHW(batch=2, channels=3, depth=5, height=7, width=6)
     filter_shape = FilterShape3D(
         depth=3, height=3, width=3, in_channels=3, out_channels=2)
     in_op = self._random_data_op(in_shape)
     filter_op = self._random_data_op(filter_shape)
-    strides = [1, 1, 1, 1, 1]
-    padding = 'VALID'
-    dilations = [1, 1, 2, 2, 2]
-    out_op = nn_ops.conv3d(
-        in_op,
-        filter_op,
-        strides=strides,
-        padding=padding,
-        data_format='NCDHW',
-        dilations=dilations)
-    self._assert_reproducible(out_op)
+    self._assert_reproducible(lambda: nn_ops.conv3d(
+        in_op, filter_op, strides=[1, 1, 1, 1, 1], padding='VALID',
+        data_format='NCDHW', dilations=[1, 1, 2, 2, 2]))
 
+  # This test intends to exercise nondeterminism of convolution in the forward
+  # direction for XLA only. cuDNN does not have nondeterministic forward
+  # convolution algorithms.
```

Comment:
This comment seems to imply that this test case tests that XLA convolutions are not deterministic even when determinism is enabled. I would phrase it something like "This test is primarily testing XLA, since cuDNN forward convolutions are always deterministic, even when determinism is not enabled. The convolution configuration tested is nondeterministic with XLA when determinism is disabled".

Also, does XLA not use cuDNN for convolutions? I am unfamiliar with the details of XLA codegen here.

Reply:
I'll clarify the comment.

Regarding XLA's use of cuDNN for convolutions: historically, XLA definitely used cuDNN for convolutions, and I didn't think that had changed (based on a recent perusal of the relevant code). However, while playing around with these tests, I happened to find that this 3D convolution case exercises nondeterminism when op-determinism is not enabled, but only for the XLA auto-JIT case. I don't understand why it's like that.